### ISSUE TITLE
Mount a FUSE filesystem without use of root or fusermount (suid)

### ISSUE THREAD

**Issue by probonopd** [AppImage/type2-runtime#71]
As @mgord9518 [pointed out](https://github.com/AppImage/type2-runtime/pull/32#issuecomment-2296448370):

https://github.com/AppImage/type2-runtime/pull/32#issuecomment-2296448370

> It’s possible to mount a FUSE filesystem without use of root permissions or SUID binaries by doing the mount inside of a user namespace.

VERY interesting @Mgord. 💯 I think you are up to something. That suid helper binary always bothered be to begin with.

Does anyone know how to actually implement this, in code? Any help appreciated 👍

**Comment by mgord9518** [comment]
For a trivial example, you can just use `unshare --mount --user -r`, which will create a fake root environment which FUSE can be mounted from. An issue I see is that it's almost like a sandbox, so the behavior won't just be 1:1 with using fusermount.

Maybe someone should experiment with launching AppImages from this shell environment to see how differently the applications behave before implementing anything in C

**Comment by TheAssassin** [comment]
> user namespace

For security reasons, user namespaces have been restricted resp. locked down more again on many OSes. I could imagine that this would cause issues with rootless FUSE. Plus, they've never been enabled on a lot of popular distros.

**Comment by probonopd** [comment]
Isn't Bubblewrap using this, too? Which would mean that at least all distributions that support Flatpak should have this enabled by default?

**Comment by mgord9518** [comment]
> Isn't Bubblewrap using this, too? Which would mean that at least all distributions that support Flatpak should have this enabled by default?

As far as I know. I believe it also has an SUID version but normal bubblewrap uses unprivileged namespaces. Even with the distros that disable them (like Ubuntu), many distros still support them out of the box. Nix, Arch and LM right off the top of my head, so assuming we can actually get it to feel like there's no difference it's probably worth it

I've started on making a new AppImage runtime in Zig and I'll try to implement mounting with namespaces

**Comment by Samueru-sama** [comment]
> For a trivial example, you can just use unshare --mount --user -r, which will create a fake root environment which FUSE can be mounted from. An issue I see is that it's almost like a sandbox, so the behavior won't just be 1:1 with using fusermount.

> Maybe someone should experiment with launching AppImages from this shell environment to see how differently the applications behave before implementing anything in C

I tested this with the nvtop appimage, it has an odd issue that I can no longer see the active processes, it also breaks the cpu-x daemon.

**Comment by dleggo** [comment]
Any news on this?

**Comment by Samueru-sama** [comment]
> Any news on this?

I don't think this will be implemented, the ideal solution for now is to set `APPIMAGE_EXTRACT_AND_RUN` when fuse isn't available. 

[see this for example](https://github.com/mgord9518/aisap/pull/20)

Would be great if the runtime actually did it without having to set the env variable though.

**Comment by TheAssassin** [comment]
> Would be great if the runtime actually did it without having to set the env variable though.

The reason it doesn't is that it's a very inefficient method and we would rather have people set up FUSE. There is a significant share of people running into this who can actually fix FUSE. The few who don't most likely know about the flag.

What I am open to is to check if FUSE does work, then check whether we're in a Docker environment (or other environments which cannot really be fixed) and then print a warning and extract the files. I was objected to that in the past, but given the fact that enabling FUSE in namespaces is a bad idea security wise, it'd be a small quality-of-life improvement to do auto extract.

For the record, the extracting could be optimized a lot to, e.g., cache more reliably.

**Comment by Samueru-sama** [comment]
Yeah I'm fine with it only being done automatically when we are on some container that doesn't allow SUID binaries.

The [uruntime](https://github.com/VHSgunzo/uruntime) actually cleans up the extracted files when the application stops being used when `APPIMAGE_EXTRACT_AND_RUN` is used while the type2-runtime leaves them on `/tmp`.

**Comment by TheAssassin** [comment]
> The [uruntime](https://github.com/VHSgunzo/uruntime) actually cleans up the extracted files when the application stops being used when APPIMAGE_EXTRACT_AND_RUN is used while the type2-runtime leaves them on /tmp.

I think I initially implemented a `$NO_CLEANUP` and had it always clean up, but I think that was changed at some point to clean up only when asked to because you might end up running two instances at the same time (they both extract/cache in the same path) and then one instance might end up deleting the other's files. There are ways around this, e.g., spinning forking off a small daemon that tracks the instances.

**Comment by dleggo** [comment]
For cases where FUSE isn't wanted it would be nice to have a user-namespaces option. Personally, to run AppImages, I extract them even though I have FUSE installed just because it is much faster, (1.5x - 2x) but user namespaces can utilize the kernel mounts which are much faster. Additionally, for some users, FUSE is a big [barier to entry](https://www.joelonsoftware.com/2000/06/03/strategy-letter-iii-let-me-go-back/), and it would be nice to eliminate that for people who have user namespaces enabled in their kernel.

**Comment by Samueru-sama** [comment]
>Personally, to run AppImages, I extract them even though I have FUSE installed just because it is much faster, (1.5x - 2x) 

In what AppImages do you see this? With the type2-runtime and zstd compression the overhead I see is usually less than 1 second for most applications. And when it is more than 1 second it is usually apps that have high launch times anyway without being appimage. 

There is room for improvement, I personally use the uruntime because it supports dwarfs, [which reduces the launch time even more](https://github.com/AppImage/AppImageSpec/issues/36#issuecomment-2438975604) and even with squashfs the uruntime is still slightly [faster](https://github.com/AppImage/type2-runtime/issues/116).

**Comment by dleggo** [comment]
> In what AppImages do you see this?

Time in seconds for applications to launch. (Not perfectly accurate, but I tried to get it as accurate as possible.)
Brave AppImage launch: 3.210
Brave Extracted: 1.960
OnlyOffice AppImage: 3.213
OnlyOffice Extracted: 1.943
VSCode AppImage: 6.754
VSCode Extracted: 6.004
Inkscape AppImage: 3.094
Inkscape Extracted: 1.994


> And when it is more than 1 second it is usually apps that have high launch times anyway without being appimage.

This kinda sounds multiplicative to me. :)

I would love to see more work in the area of user namespaces as they have the opportunity to speed up many things that are currently using FUSE!  Performance is actually a relatively common complaint against AppImage IIRC, and this would help migate that.

**Comment by TheAssassin** [comment]
Have you used the flag or pre-extracted the AppImages?

**Comment by dleggo** [comment]
Pre-extracted

**Comment by Samueru-sama** [comment]
> Extracted: 1.960
> OnlyOffice AppImage: 3.213
> OnlyOffice Extracted: 1.943
> VSCode AppImage: 6.754

Brave actually uses the static runtime, I made the PR adding that a while ago lol. 

![image](https://github.com/user-attachments/assets/3e97645a-28c3-4e35-a6f5-1cfdb64e1719)

Welp, it was more than 1 second 🤣 It would be if the uruntime was being used since it doesn't have the [25% degradation](https://github.com/AppImage/type2-runtime/issues/116) from the static musl. 

Before thedev switched to linking glibc statically, the uruntime was actually slower than the type2-runtime, if the type2-runtime has the same issue the fix would bring the time down to ~1.5 seconds in this example further reducing the overhead. 

----------------------------------------------------------------------------

Inkscape is an interesting case, they actually use go-appimage to deploy the appimage and bundle all the libraries, meaning this is one of the few appimages that would work on musl systems if it wasn't for the fact that they are still using the old dynamic runtime. 

The developer is still using the old appimagetool from appimagekit because of issues with go-appimage appstream verification, this can actually be fixed by switching to the updated appimagetool and iirc I already mentioned that but was ignored it seems lol

Their AppImage uses gzip compression right now which is slower:

![image](https://github.com/user-attachments/assets/53be990d-e191-4b3a-9b97-d65d119401b7)

Takes ~2.5 seconds while zstd level 15 takes ~1.9 seconds. 

And for reference the extracted AppImage takes 1.2 seconds.

----------------------------------------------------------------------------

Anyways don't get me wrong, if namespaces are much faster then that change is more than welcomed, but I need to see/do some tests, is it enough to do the unshare trick that was suggested here before to test or that doesn't count?

**Comment by Samueru-sama** [comment]
Just tested with unshare anyway. Made two wrapper scripts since I have to pass the `--no-sandbox` in this case because otherwise Brave won't open and the script that I use to benchmark only takes the window class as an argument:

![image](https://github.com/user-attachments/assets/9cb2af1b-b6c9-49fe-9403-f311feb535f5)

👀 well it is actually ~20% faster indeed.

**Comment by dleggo** [comment]
I am not actually certain how big the preformance gap is between FUSE and kernel mounting. ( Could be of reference https://github.com/apptainer/apptainer/issues/665 ) Although it doesn't appear very significant. 

Those tests you did have quite interesting results though. That is still with FUSE but just inside a user namespace? I actually don't know how that made it faster :laughing: . I was more advocating for the usage of kernel mounting inside the namespace, but know I see I misread probonopd's post and the title! Whoops! :laughing: It is quite fascinating that FUSE inside a user namespace is faster though... Sorry for my probably irrelevent posts.

**Comment by Samueru-sama** [comment]
I also don't know how it is faster, would be nice if anyone could try to replicate what I did though.

**Comment by dleggo** [comment]
The method I got my rough numbers was to run
`time ./Brave-*.AppImage` and Alt+F4 it the moment content in the window appeared, so it was much less scientific then the tests you did. :laughing: Would you mind posting the contents of your `benchmark-startup.sh` script and I could test of I can replicate the results, if you want?

**Comment by Samueru-sama** [comment]
> The method I got my rough numbers was to run `time ./Brave-*.AppImage` and Alt+F4 it the moment content in the window appeared, so it was much less scientific then the tests you did. 😆 Would you mind posting the contents of your `benchmark-startup.sh` script and I could test of I can replicate the results, if you want?

```
#!/bin/bash

# This script is used to determines the startup time of applications
# It only works if your WM is configured to focus on new windows automatically

# Start the application
"$1" >/dev/null 2>&1 &

# START TIME (nanoseconds)
START=$(date +%s%N)

# Loop until the window class is found by i3
while true; do
    window_class="$(xdotool getactivewindow getwindowname 2>/dev/null)"
    if echo "$window_class" | grep -qi "$2"; then
        break
    fi
    sleep 0.001
done

# END TIME
END=$(date +%s%N)
TIME=$((($END - $START) / 1000000))
echo "Time taken: $TIME miliseconds"
```

All it does is launch the app and count until the window name matches `$2` which I just use brave in this case. 

This script will only work if you are on X11 and you have your window manager configured to focus on new windows automatically.

**Comment by dleggo** [comment]
Thanks for the script! Unfortunately I use Wayland... I did 10 tests with my previous method inside and outside user namespaces with Inkscape, and there was no difference. (2.5775s outside namespace, 2.5636s inside namespace) It definitely isn't slower though. I can log into an X11 session if needed though?

**Comment by Samueru-sama** [comment]
![image](https://github.com/user-attachments/assets/fb2e966f-12ad-44a4-a1f1-ce5dd2f6d099)

Seems there is no difference with inkscape, but I did have two instances where the time was slightly lower when using unshare 🤔

There is a diference with Brave though: 

![image](https://github.com/user-attachments/assets/64fe51c0-c557-4f76-8b28-1e81544c4e2f)

**Comment by dleggo** [comment]
Interesting... If it doesn't appear to have any performance downsides then perhaps this could be useful for the sole purpose of not having to have a SUID helper binary?

> An issue I see is that it's almost like a sandbox, so the behavior won't just be 1:1 with using fusermount.

Unfortunately, user namespaces need to operate this way. I don't believe there is a  means to allow a process in a different namespace to see outside processes IIRC.

**Comment by Samueru-sama** [comment]
> Interesting... If it doesn't appear to have any performance downsides then perhaps this could be useful for the sole purpose of not having to have a SUID helper binary?
> 
> > I tested this with the nvtop appimage, it has an odd issue that I can no longer see the active processes, it also breaks the cpu-x daemon.
> 
> After some testing, i.e.
> 
> ```
> $ unshare -Urm
> # ps aux
> # top
> ```
> 
> I can still see all outside processes... Are you accidentally unsharing the PID namespace @Samueru-sama ?

I originally just used the `unshare --mount --user -r ./nvtop.AppImage` and that results in not being able to see processes. 

I tried `unshare -Urm` and the nvtop and I get the same issue: 

With unshare

![image](https://github.com/user-attachments/assets/60738493-5161-4a9c-8498-a13fcee3d3f5)

Without unshare

![image](https://github.com/user-attachments/assets/e8e3689e-bfbe-4924-8bd8-514b0dd79f74)

**Comment by dleggo** [comment]
I tried it myself and it had the same behavior, quite peculiar... I tested `top` `htop` `ps aux` and they show the correct results. I tested both the AppImage and non-AppImage and they both failed to function, so maybe it is a bug with nvtop then? Maybe it reads the process list in a different way? I tried `strace nvtop` inside the namespace and came across some permission issues, but after trying `sudo unshare -Um` nothing changed, so my guess is that this is some strange nvtop behaviour.

>I originally just used the unshare --mount --user -r ./nvtop.AppImage and that results in not being able to see processes.
>
>I tried unshare -Urm and then nvtop and I get the same issue:

` unshare --mount --user -r` and `unshare -Urm` are actually the same command, I am just used to using the abbreviated version. :)

**Comment by xplshn** [comment]
![Image](https://github.com/user-attachments/assets/1484f6d4-c8cf-4acc-8807-bb1b5f9e32a5)
Using unshare didn't work for me either

**Comment by probonopd** [comment]
Please note that AppBundle files do not use the type2-runtime from this repo.

**Comment by xplshn** [comment]
> Please note that AppBundle files do not use the type2-runtime from this repo.

Yeah. Its related, we're all facing the same issue.

**Comment by dleggo** [comment]
Are you using the `unshare` command specified at the beginning of this thread? i.e. `unshare -Urm`

**Comment by dleggo** [comment]
@xplshn, I cannot be certain exactly the issue, but I do notice that your AppBundle is trying to use fusermount? Fusermount shouldn't be necessary when inside a user namespace (confirmed by testing AppImages in an environment without fusermount). If this is the problem you are running into then this is not related to AppImage, but this could be discussed in a pElf issue?

**Comment by xplshn** [comment]
> [@xplshn](https://github.com/xplshn), I cannot be certain exactly the issue, but I do notice that your AppBundle is trying to use fusermount? Fusermount shouldn't be necessary when inside a user namespace (confirmed by testing AppImages in an environment without fusermount). If this is the problem you are running into then this is not related to AppImage, but this could be discussed in a pElf issue?

I made a wrapper for fusermount and fusermount3, the wrapper works and uses unshare. But it doesn't work when `dwarfs`/`squashfuse` calls it

**Comment by dleggo** [comment]
Maybe lets continue this discussion at https://github.com/xplshn/pelf/issues/122 as this is off topic for this thread. :smiley:

**Comment by dleggo** [comment]
The biggest problem with using the user-namespace approach is that it changes the apparent user ID. Anybody know of a way to rectify this while still not requiring fusermount?

**Comment by mgord9518** [comment]
@dleggo Can you not run su in a namespace? I'm away from my PC at the moment so I can't test it but that may be a simple solution if it works

**Comment by dleggo** [comment]
@mgord9518 , no you can't unfortunately. It will throw

```
su: Authentication service cannot retrieve authentication info
```

**Comment by xplshn** [comment]
> [@mgord9518](https://github.com/mgord9518) , no you can't unfortunately. It will throw
> 
> ```
> su: Authentication service cannot retrieve authentication info
> ```

Without PAM the error message changes to:
```
]~/Documents/TrulyMine/ocik@ unshare -Urm su
su: can't set groups: Operation not permitted
]~/Documents/TrulyMine/ocik@ 
```

**Comment by dleggo** [comment]
Unfortunately it doesn't seem like there is a way around it, unless if anyone knows a way to fake a user ID? Something like `fakeroot` that fakes a normal user account instead?

**Comment by xplshn** [comment]
> Unfortunately it doesn't seem like there is a way around it, unless if anyone knows a way to fake a user ID? Something like `fakeroot` that fakes a normal user account instead?

`bwrap --uid 0 --gid 0`

**Comment by dleggo** [comment]
I don't believe fuse will work inside an environment where it is not mapped to root IIRC.

**Comment by xplshn** [comment]
> That creates a new namespace, which unfortunately doesn't work for this case. Good idea though! 😀

![Image](https://github.com/user-attachments/assets/6cad1d54-596e-403b-bb17-c971ea1a2cb0)

Exactly like bwrap

**Comment by dleggo** [comment]
Sorry, I understood that unshare did that, I was meaning _another_ namespace. Though I wasn't aware one could create nested namespaces while still retaining mount points! After further looking, one can mount the fuse filesystem, and then use bubblewrap to create a nested namespace using the UID and GID options you specified above. It seems to me like this could be a reasonable route?
